src/main/java/com/thebois/models/beings/roles/AbstractHarvesterRole.java

```diff
@@ -0,0 +1,74 @@
+package com.thebois.models.beings.roles;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+
+import com.thebois.abstractions.IResourceFinder;
+import com.thebois.models.Position;
+import com.thebois.models.beings.IActionPerformer;
+import com.thebois.models.beings.actions.ActionFactory;
+import com.thebois.models.beings.actions.IAction;
+import com.thebois.models.beings.actions.IActionGenerator;
+import com.thebois.models.world.IWorld;
+import com.thebois.models.world.resources.IResource;
+import com.thebois.models.world.resources.ResourceType;
+
+/**
+ * Represents a role that mainly harvests items from resources.
+ */
+abstract class AbstractHarvesterRole extends AbstractRole {
+
+    private final IResourceFinder finder;
+    private final IWorld world;
+    private final ResourceType resourceType;
+
+    /**
+     * Instantiates with a way of finding resources.
+     *
+     * @param finder       The locator of resources.
+     * @param world        The world in which the resources are located.
+     * @param resourceType The type of resource to gather.
+     */
+    AbstractHarvesterRole(
+        final IResourceFinder finder, final IWorld world, final ResourceType resourceType) {
+        this.finder = finder;
+        this.world = world;
+        this.resourceType = resourceType;
+    }
+
+    @Override
+    protected Collection<IActionGenerator> getTaskGenerators() {
+        return List.of(this::createMoveToResource, this::createHarvestResource);
+    }
+
+    protected IAction createMoveToResource(final IActionPerformer performer) {
+        final Optional<IResource> maybeResource = findNearbyResource(performer);
+        if (maybeResource.isEmpty()) return ActionFactory.createDoNothing();
+
+        final IResource resource = maybeResource.get();
+
+        final Position position = performer.getPosition();
+        final Optional<Position> closestSpotNextToResource = world.getClosestNeighbourOf(resource,
+                                                                                         position);
+
+        if (closestSpotNextToResource.isEmpty()) return ActionFactory.createDoNothing();
+
+        return ActionFactory.createMoveTo(closestSpotNextToResource.get());
+    }
+
+    protected IAction createHarvestResource(final IActionPerformer performer) {
```

Comment:
I might have missed something, but why are these methods protected. The only method I can see being used by subclasses are getTaskGenerators but not the other 3 methods.

Reply:
You're right, they're not used in any subtype as of now.

I made them protected because I can imagine that some harvesting-roles will want to define different ways of finding resources, and different ways of harvesting them. But maybe that assumption is wrong, what do you think?

The argument could be made that this is speculative generality, and that they should only be made protected when they actually need to be, not before.

Reply:
The downside to making them private now is that whenever some hypothetical future programmer wants to implement a unique harvesting role that finds resources only by wandering around (instead of instantly knowing where they are) and that only harvests them by dancing around in a circle first, they'll have to go way up into `AbstractHarvesterRole` and change the visibility of these methods.

My reasoning is that, since it is an abstract class, it's good that it declares some methods as protected, so subtypes can access/override them.

Reply:
Hmm, I can understand your way of thinking, but I still feel like they should be private until they need to be protected. 
What do you think @jonathan-lindqvist  and @JacobBredin?

Reply:
I agree with @Matte02 
They should be private, taking the _hypothetical future programmer_ argument to the unnecessary extreme, we could make everything public because someone in the future might want to access this.